shared/src/server/health.rs

```diff
@@ -55,3 +71,132 @@ impl std::fmt::Display for HealthCheckStatus {
         }
     }
 }
+
+pub trait HealthCheck {
+    fn status_code(&self) -> u16;
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum ControlPlaneState {
+    Draining,
+    Ok,
+}
+
+impl HealthCheck for ControlPlaneState {
+    fn status_code(&self) -> u16 {
+        match self {
+            ControlPlaneState::Ok => 200,
+            ControlPlaneState::Draining => 500,
+        }
+    }
+}
+
+impl HealthCheck for DataPlaneState {
+    fn status_code(&self) -> u16 {
+        match self {
+            DataPlaneState::Initialized(diagnostic) if diagnostic.is_healthy() => 200,
+            _ => 500,
+        }
+    }
+}
+
+impl HealthCheck for HealthCheckLog {
+    fn status_code(&self) -> u16 {
+        self.status_code()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum DataPlaneState {
+    Error(String),
+    Unknown(String),
+    Provisioning,
+    Attesting,
+    SourcingTlsCerts,
+    Initialized(DataPlaneDiagnostic),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct DataPlaneDiagnostic {
+    pub user_process: UserProcessHealth,
+}
+
+impl DataPlaneDiagnostic {
+    pub fn is_healthy(&self) -> bool {
+        match self.user_process {
+            UserProcessHealth::Unknown(_) => true,
+            UserProcessHealth::Error(_) => false,
+            UserProcessHealth::Response { status_code, .. } => status_code == 200,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum UserProcessHealth {
+    Unknown(String),
+    Error(String),
+    Response {
+        status_code: u16,
+        body: Option<serde_json::Value>,
+    },
+}
+
+impl UserProcessHealth {
+    pub fn rank(&self) -> u8 {
+        match self {
+            UserProcessHealth::Unknown(_) => 0,
+            UserProcessHealth::Error(_) => 1,
+            UserProcessHealth::Response { .. } => 2,
```

Comment:
don't think this will work... this could bury an unhealthy result? Considering something like (most recent to least recent):
```rust
[Error(Socket Hang Up), Response(200)]
```

we would take response as max and drop error?

Reply:
May be misunderstanding how we're using this.

Reply:
Yea the idea is we should always consider UP responses first. Because it could also be
```rust
[Error(Socket Hang Up), Response(500)]
```
If we're getting socket hangups (ie the UP is dead?) then the healthcheck agent's buffer should easily become full within 5 seconds.

I guess there could be a case where we're consistently getting the vec you provided for a long period of time and that could cause an issue? I'm not sure how possible it would be to continue collecting UP healthchecks alongside errors occurring performing the healthcheck?

Reply:
would it be more accurate to rank responses according to status code of response?
```rust
Unknown = 0
Error => 2,
Response { status_code, .. } => if status_code.is_ok() { 1 } else { 3 } 
```
So we always prioritise unhealthy responses over healthy, but for user vis an API response with an unhealthy status code takes priority over a generic error